File: tests/solutions_cpp/number_of_connected_components.cpp
#include "json_helper.h"

class Solution {
public:
    int countComponents(int n, vector<vector<int>>& edges) {
        vector<int> parent(n);
        iota(parent.begin(), parent.end(), 0);
        vector<int> rank_(n, 0);
        function<int(int)> find = [&](int x) {
            while (parent[x] != x) {
                parent[x] = parent[parent[x]];
                x = parent[x];
            }
            return x;
        };
        int components = n;
        for (auto& e : edges) {
            int a = find(e[0]), b = find(e[1]);
            if (a == b) continue;
            if (rank_[a] < rank_[b]) swap(a, b);
            parent[b] = a;
            if (rank_[a] == rank_[b]) rank_[a]++;
            components--;
        }
        return components;
    }
};

int main() {
    auto args = parse_args();
    Solution sol;
    int n = to_int(args[0]);
    auto edges = to_vector_vector_int(args[1]);
    cout << sol.countComponents(n, edges) << endl;
    return 0;
}

File: tests/solutions_js/number_of_connected_components.js
class Solution {
    countComponents(n, edges) {
        const parent = Array.from({ length: n }, (_, i) => i);
        const rank = new Array(n).fill(0);

        const find = (x) => {
            if (parent[x] !== x) {
                parent[x] = find(parent[x]);
            }
            return parent[x];
        };

        const union = (x, y) => {
            let px = find(x);
            let py = find(y);
            if (px === py) return 0;
            if (rank[px] < rank[py]) [px, py] = [py, px];
            parent[py] = px;
            if (rank[px] === rank[py]) rank[px]++;
            return 1;
        };

        let components = n;
        for (const [u, v] of edges) {
            components -= union(u, v);
        }
        return components;
    }
}
